### ISSUE TITLE
systemd-networkd DHCP Server ignores SendOptions

### ISSUE THREAD

**Issue by benibr** [systemd/systemd#15780]
**systemd version the issue has been seen with**
245 (245.4-4ubuntu3)

**Used distribution**
VERSION="20.04 LTS (Focal Fossa)"

didn't work in Ubuntu 18.04 either

**Expected behaviour you didn't see**
If I config a DHCP server with systemd-networkd I want to add additional `SendOptions` to tell the client about PXE boot servers and files.

**Unexpected behaviour you saw**
However systemd-networkd ignores all these options and sends only implicit configured options like gateway/DNS etc. but none of the explicit defined `SendOptions`

**Steps to reproduce the problem**
***config files***
```/etc/systemd/network/10-pxe.link
[Match]
MACAddress=00:11:22:33:44:55

[Link]
Name=pxe
```

```/etc/systemd/network/10-lower.network
[Match]
Name=pxe

[Network]
Description="Client installation network with DHCP server and PXE boot"
DHCP=no
DHCPServer=yes
Address=10.23.23.254/24

[DHCPServer]
PoolOffset=100
PoolSize=25
SendOption=128:ipv4address:10.23.23.10
SendOption=67:string:ipxe.efi
```

***trying dhcp***
(Re-)start systemd-networkd and for example bridge an EFI VM to it which tries network boot.
TCP Dump on the systemd-networkd machine shows the following:

```
17:06:46.297872 IP (tos 0x0, ttl 64, id 51369, offset 0, flags [none], proto UDP (17), length 375)
    0.0.0.0.68 > 255.255.255.255.67: BOOTP/DHCP, Request from 52:54:00:f8:35:bc, length 347, xid 0x4a1d1716, Flags [Broadcast]
          Client-Ethernet-Address 52:54:00:f8:35:bc
          Vendor-rfc1048 Extensions
            Magic Cookie 0x63825363
            DHCP-Message Option 53, length 1: Discover
            MSZ Option 57, length 2: 1472
            Parameter-Request Option 55, length 35:
              Subnet-Mask, Time-Zone, Default-Gateway, Time-Server
              IEN-Name-Server, Domain-Name-Server, Hostname, BS
              Domain-Name, RP, EP, RSZ
              TTL, BR, YD, YS
              NTP, Vendor-Option, Requested-IP, Lease-Time
              Server-ID, RN, RB, Vendor-Class
              TFTP, BF, GUID, Option 128
              Option 129, Option 130, Option 131, Option 132
              Option 133, Option 134, Option 135
            GUID Option 97, length 17: 0.238.188.149.198.210.123.74.66.148.204.160.14.125.242.5.148
            NDI Option 94, length 3: 1.3.1
            ARCH Option 93, length 2: 7
            Vendor-Class Option 60, length 32: "PXEClient:Arch:00007:UNDI:003001"
17:06:46.298248 IP (tos 0xc0, ttl 64, id 55807, offset 0, flags [DF], proto UDP (17), length 296)
    10.23.23.254.67 > 255.255.255.255.68: BOOTP/DHCP, Reply, length 268, xid 0x4a1d1716, Flags [Broadcast]
          Your-IP 10.23.23.123
          Client-Ethernet-Address 52:54:00:f8:35:bc
          Vendor-rfc1048 Extensions
            Magic Cookie 0x63825363
            DHCP-Message Option 53, length 1: Offer
            Lease-Time Option 51, length 4: 3600
            Subnet-Mask Option 1, length 4: 255.255.255.0
            Default-Gateway Option 3, length 4: 10.23.23.254
            Server-ID Option 54, length 4: 10.23.23.254
```

As you can see the client asks for TFTP and some other PXE related options but the answer does not contain them.
I tried different clients (hardware networkcards, machines with BIOS/UEFI, VMs) the all keep asking for a while and then give up.

I also tried other options (eg. 128-135) and they are all ignored. The same applies to SendOption values within "

Any suggestions or help appreciated :-)

**Comment by benibr** [comment]
For reference I provide a dnsmasq config which works as expected:

```/etc/dnsmasq.conf
interface=pxe
except-interface=lo
bind-interfaces
domain=example.com
# Set default gateway
dhcp-option=3,10.23.23.254
# Set DNS servers to announce
dhcp-option=6,10.23.23.254
dhcp-option=67,ipxe.efi
dhcp-option=150,10.23.23.254

dhcp-range=10.23.23.10,10.23.23.20,12h
```

```
17:17:31.563188 IP (tos 0x0, ttl 64, id 7352, offset 0, flags [none], proto UDP (17), length 375)
    0.0.0.0.68 > 255.255.255.255.67: BOOTP/DHCP, Request from 52:54:00:f8:35:bc, length 347, xid 0xb9a36b25, Flags [Broadcast]
          Client-Ethernet-Address 52:54:00:f8:35:bc
          Vendor-rfc1048 Extensions
            Magic Cookie 0x63825363
            DHCP-Message Option 53, length 1: Discover
            MSZ Option 57, length 2: 1472
            Parameter-Request Option 55, length 35:
              Subnet-Mask, Time-Zone, Default-Gateway, Time-Server
              IEN-Name-Server, Domain-Name-Server, Hostname, BS
              Domain-Name, RP, EP, RSZ
              TTL, BR, YD, YS
              NTP, Vendor-Option, Requested-IP, Lease-Time
              Server-ID, RN, RB, Vendor-Class
              TFTP, BF, GUID, Option 128
              Option 129, Option 130, Option 131, Option 132
              Option 133, Option 134, Option 135
            GUID Option 97, length 17: 0.238.188.149.198.210.123.74.66.148.204.160.14.125.242.5.148
            NDI Option 94, length 3: 1.3.1
            ARCH Option 93, length 2: 7
            Vendor-Class Option 60, length 32: "PXEClient:Arch:00007:UNDI:003001"
17:17:31.564242 IP (tos 0xc0, ttl 64, id 5029, offset 0, flags [none], proto UDP (17), length 347)
    10.23.23.254.67 > 255.255.255.255.68: BOOTP/DHCP, Reply, length 319, xid 0xb9a36b25, Flags [Broadcast]
          Your-IP 10.23.23.20
          Server-IP 10.23.23.254
          Client-Ethernet-Address 52:54:00:f8:35:bc
          Vendor-rfc1048 Extensions
            Magic Cookie 0x63825363
            DHCP-Message Option 53, length 1: Offer
            Server-ID Option 54, length 4: 10.23.23.254
            Lease-Time Option 51, length 4: 43200
            RN Option 58, length 4: 21600
            RB Option 59, length 4: 37800
            Subnet-Mask Option 1, length 4: 255.255.255.0
            BR Option 28, length 4: 10.23.23.255
            Domain-Name Option 15, length 14: "example.com"
            BF Option 67, length 9: "ipxe.efi^@"
            Domain-Name-Server Option 6, length 4: 10.23.23.254
            Default-Gateway Option 3, length 4: 10.23.23.254
```

**Comment by ssahani** [comment]
I tested with 

```

[DHCPServer]
PoolOffset=10
PoolSize=50
POP3Servers=192.168.5.1
SendOption=128:ipv4address:10.23.23.10
SendOption=67:string:ipxe.efi
SendOption=129:string:hello1
SendOption=130:string:hello2
SendOption=131:string:hello3
SendOption=132:string:hello4
SendOption=133:string:hello5
SendOption=135:string:hello6
```

```
Frame 62: 395 bytes on wire (3160 bits), 395 bytes captured (3160 bits) on interface veth99, id 0
Ethernet II, Src: b6:59:12:c8:81:93 (b6:59:12:c8:81:93), Dst: 1e:04:f8:b8:2f:d4 (1e:04:f8:b8:2f:d4)
Internet Protocol Version 4, Src: 192.168.5.1, Dst: 192.168.5.35
User Datagram Protocol, Src Port: 67, Dst Port: 68
Dynamic Host Configuration Protocol (ACK)
    Message type: Boot Reply (2)
    Hardware type: Ethernet (0x01)
    Hardware address length: 6
    Hops: 0
    Transaction ID: 0x8748f1a7
    Seconds elapsed: 0
    Bootp flags: 0x0000 (Unicast)
    Client IP address: 0.0.0.0
    Your (client) IP address: 192.168.5.35
    Next server IP address: 0.0.0.0
    Relay agent IP address: 0.0.0.0
    Client MAC address: 1e:04:f8:b8:2f:d4 (1e:04:f8:b8:2f:d4)
    Client hardware address padding: 00000000000000000000
    Server host name not given
    Boot file name not given
    Magic cookie: DHCP
    Option: (53) DHCP Message Type (ACK)
        Length: 1
        DHCP: ACK (5)
    Option: (51) IP Address Lease Time
        Length: 4
        IP Address Lease Time: (3600s) 1 hour
    Option: (1) Subnet Mask (255.255.255.0)
        Length: 4
        Subnet Mask: 255.255.255.0
    Option: (3) Router
        Length: 4
        Router: 192.168.5.1
    Option: (70) POP3 Server
        Length: 4
        POP3 Server: 192.168.5.1
    Option: (101) TCode
        Length: 13
        TZ TCode: Europe/Berlin
    Option: (128) DOCSIS full security server IP [TODO]
        Length: 4
        Value: 0a17170a
    Option: (67) Bootfile name
        Length: 8
        Bootfile name: ipxe.efi
    Option: (129) PXE - undefined (vendor specific)
        Length: 6
        Value: 68656c6c6f31
    Option: (130) PXE - undefined (vendor specific)
        Length: 6
        Value: 68656c6c6f32
    Option: (131) PXE - undefined (vendor specific)
        Length: 6
        Value: 68656c6c6f33
    Option: (132) PXE - undefined (vendor specific)
        Length: 6
        Value: 68656c6c6f34
    Option: (133) PXE - undefined (vendor specific)
        Length: 6
        Value: 68656c6c6f35
    Option: (135) PXE - undefined (vendor specific)
        Length: 6
        Value: 68656c6c6f36
    Option: (54) DHCP Server Identifier (192.168.5.1)
        Length: 4
        DHCP Server Identifier: 192.168.5.1
    Option: (255) End
        Option End: 255

```
```

**Comment by grillo-parlante** [comment]
@ssahani: In the traces that @benibr gave, he was looking at DHCP OFFER packets, but your trace is showing an ACK packet.  In other words, PXE/BOOTP requires (at least) the ability to specify option 66 and 67 for **OFFER** packets.  (I am uncertain if DHCP requires that other options be sent with ACK packets.)

Looking at [src/libsystemd-network/sd-dhcp-server.c](https://github.com/systemd/systemd/blob/988338eb19ee2049611da3a8bbce9ff4525e5c61/src/libsystemd-network/sd-dhcp-server.c#L428) confirms that the `ORDERED_SET_FOREACH(j, server->extra_options) { … }` which appends user-specified options to the packet, is absent from server_send_offer() but is present in server_send_ack().

**Comment by yuwata** [comment]
@grillo-parlante Thank you for the investigation. Fix is waiting in #18874.
